### PR TITLE
io: add option to overwrite Event/Execute

### PR DESCRIPTION
**Pull request by inkeliz**
This change makes possible to change the default
behaviour, and also expose Router.Execute.

### CODE REVIEW

**Comment by ddkwork** [comment]
Is it possible to skip interfaces and global variables in favor of generics

**Comment by inkeliz** [comment]
Sorry, but can you clarify how would Generics be implemented here?


The usage of generics will also requires a specific implementation, for instance in `app.FrameEvent`, `layout.Context`, which uses `Source`, or `input.Source` itself. If it's `layout.Context[T]` (then `layout.Context[Source]`) the issue still the same, since it's not possible to use other type. If `input.Source` becomes `input.Source[T any]` then we have the exactly same issue.

I don't see how Generics can be implemented here, but might missing something (I just awakup few minutes ago )

**Comment by ddkwork** [comment]
I'm currently busy refactoring another library, and I'll take the time to read the source code and analyze the feasibility of generalization soon!



---Original---
From: ***@***.***&gt;
Date: Sat, Jun 1, 2024 15:45 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [gioui/gio] io: add option to overwrite Event/Execute (PR #136)




 
Sorry, but can you clarify how would Generics be implemented here?
 
The usage of generics will also requires a specific implementation, for instance in app.FrameEvent, layout.Context, which uses Source, or input.Source itself. If it's layout.Context[T] (then layout.Context[Source]) the issue still the same, since it's not possible to use other type. If input.Source becomes input.Source[T any] then we have the exactly same issue.
 
I don't see how Generics can be implemented here, but might missing something (I just awakup few minutes ago )

File: io/input/router.go [hunk 1]
```diff
@@ -193,12 +185,34 @@ func (s Source) Focused(tag event.Tag) bool {
...
+// implement a custom event processor or testing.
+var SourceEventProcessor = (*Router).Event
+
 // Event returns the next event that matches at least one of filters.
 func (s Source) Event(filters ...event.Filter) (event.Event, bool) {
 	if !s.Enabled() {
 		return nil, false
 	}
-	return s.r.Event(filters...)
+	return SourceEventProcessor(s.r, filters...)
```
Comment (eliasnaur) [inline io/input/router.go:215]: I believe this indirection will cause `filters` to escape the heap, causing an allocation for every call to `Source.Event`.

Comment (ddkwork) [reply io/input/router.go:215]: > I believe this indirection will cause `filters` to escape the heap, causing an allocation for every call to `Source.Event`.

So a callback function method should be added?

Comment (ddkwork) [reply io/input/router.go:215]: I also prepared a file drag and drop implementation for gio for the Windows platform, but the issue of passing events is a bit confusing, as he said, the current gui can only receive events can not generate custom events which brings confusion to the implementation of the file drag and drop events

**Pull request closed without merging**